core-bundle/src/Filesystem/Filesystem.php

```diff
@@ -0,0 +1,416 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Filesystem;
+
+use Contao\StringUtil;
+use Contao\Validator;
+use League\Flysystem\Config;
+use League\Flysystem\DirectoryAttributes;
+use League\Flysystem\DirectoryListing;
+use League\Flysystem\FileAttributes;
+use League\Flysystem\FilesystemAdapter;
+use League\Flysystem\FilesystemOperator;
+use League\Flysystem\InvalidStreamProvided;
+use League\Flysystem\PathNormalizer;
+use League\Flysystem\StorageAttributes;
+use League\Flysystem\WhitespacePathNormalizer;
+use ProxyManager\Factory\LazyLoadingGhostFactory;
+use ProxyManager\Proxy\GhostObjectInterface;
+
+/**
+ * This Flysystem operator wraps a DBAFS layer around the underlying filesystem
+ * adaptor: Listing content will only access the database, operations that can
+ * mutate the filesystem, will automatically trigger a synchronization.
+ *
+ * Additionally, there is support to directly retrieve the 'extra metadata'
+ * from a file's / directory's StorageAttributes via a dedicated method.
+ *
+ * We widen the FilesystemOperator interface - instead of calling methods with
+ * resource paths (string), you can also pass in an ID (int) or UUID of a
+ * synchronized DBAFS resource.
+ *
+ * @phpstan-import-type ExtraMetadata from Dbafs
+ */
+final class Filesystem implements FilesystemOperator
+{
+    private FilesystemAdapter $adapter;
+    private Config $config;
+    private PathNormalizer $pathNormalizer;
+    private Dbafs $dbafs;
+    private LazyLoadingGhostFactory $proxyFactory;
+
+    public function __construct(Dbafs $dbafs, FilesystemAdapter $adapter, LazyLoadingGhostFactory $proxyFactory = null, Config $config = null, PathNormalizer $pathNormalizer = null)
+    {
+        $this->dbafs = $dbafs;
+        $this->adapter = $adapter;
+        $this->proxyFactory = $proxyFactory ?? new LazyLoadingGhostFactory();
+        $this->config = $config ?? new Config();
+        $this->pathNormalizer = $pathNormalizer ?? new WhitespacePathNormalizer();
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @param string|int $location
+     */
+    public function fileExists($location): bool
```

Comment:
All the methods should allow to force ignoring the Dbafs if they want to. It should not be the default though.

Reply:
:+1: already added to the todos :slightly_smiling_face:

Reply:
implemented in e21aa8064f352740870ee50415f3d3d37121d29a